tests/components/application-connector/test/compass-runtime-agent/testkit/compassruntimeagentinit/deployment.go

```diff
@@ -0,0 +1,157 @@
+	"context"
+	"fmt"
+	"github.com/avast/retry-go"
+	"github.com/hashicorp/go-multierror"
+	"github.com/kyma-project/kyma/tests/components/application-connector/test/compass-runtime-agent/testkit/compassruntimeagentinit/types"
+	v12 "k8s.io/api/apps/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v13 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	"time"
+)
+
+const (
+	CRAContainerNumber         = 0
+	ConfigurationSecretEnvName = "APP_AGENT_CONFIGURATION_SECRET"
+	CASecretEnvName            = "APP_CA_CERTIFICATES_SECRET"
+	ClusterCertSecretEnvName   = "APP_CLUSTER_CERTIFICATES_SECRET"
+)
+
+type deploymentConfiguration struct {
+	kubernetesInterface kubernetes.Interface
+	deploymentName      string
+	namespaceName       string
+}
+
+func NewDeploymentConfiguration(kubernetesInterface kubernetes.Interface, deploymentName, namespaceName string) deploymentConfiguration {
+	return deploymentConfiguration{
+		kubernetesInterface: kubernetesInterface,
+		deploymentName:      deploymentName,
+		namespaceName:       namespaceName,
+	}
+}
+
+func (dc deploymentConfiguration) Do(newCANamespacedSecretName, newClusterNamespacedCertSecretName, newConfigNamespacedSecretName string) (types.RollbackFunc, error) {
+	deploymentInterface := dc.kubernetesInterface.AppsV1().Deployments(dc.namespaceName)
+
+	deployment, err := retryGetDeployment(dc.deploymentName, deploymentInterface)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(deployment.Spec.Template.Spec.Containers) < 1 {
+		return nil, fmt.Errorf("no containers found in %s/%s deployment", "compass-system", dc.deploymentName)
+	}
+
+	previousConfigSecretNamespacedName, found := replaceEnvValue(deployment, ConfigurationSecretEnvName, newConfigNamespacedSecretName)
+	if !found {
+		return nil, fmt.Errorf("environment variable '%s' not found in %s deployment", ConfigurationSecretEnvName, dc.deploymentName)
+	}
+
+	previousCASecretNamespacedName, found := replaceEnvValue(deployment, CASecretEnvName, newCANamespacedSecretName)
+	if !found {
+		return nil, fmt.Errorf("environment variable '%s' not found in %s deployment", CASecretEnvName, dc.deploymentName)
+	}
+
+	previousCertSecretNamespacedName, found := replaceEnvValue(deployment, ClusterCertSecretEnvName, newClusterNamespacedCertSecretName)
+	if !found {
+		return nil, fmt.Errorf("environment variable '%s' not found in %s deployment", ClusterCertSecretEnvName, dc.deploymentName)
+	}
+
+	err = retryUpdateDeployment(deployment, deploymentInterface)
+	if err != nil {
+		return nil, err
+	}
+	rollbackDeploymentFunc := newRollbackDeploymentFunc(dc.deploymentName, previousConfigSecretNamespacedName, previousCASecretNamespacedName, previousCertSecretNamespacedName, deploymentInterface)
+
+	err = waitForRollout(dc.deploymentName, deploymentInterface)
+	if err != nil {
+		var murliErr *multierror.Error
+
+		multierror.Append(murliErr, err)
+
+		err := rollbackDeploymentFunc()
+		if err != nil {
+			multierror.Append(murliErr, err)
+		}
```

Comment:
And also there. Why the rollback function is called there already? Shouldn't it be some higher where all the rollback functions would be called?

Reply:
There is a similar case as in Compass Configurator. I will remove this code.